fourier/fourier.go

```diff
@@ -0,0 +1,254 @@
+	}
+	return dst
+}
+
+// IFFT computes the real perodic sequence from the Fourier coefficients,
+// coeff, placing the result in dst and returning it. This transform is
+// unnormalized since a call to FFT followed by a call of IFFT will
+// multiply the input sequence by the length of the sequence.
+//
+// If the length of coeff is not t.Len()/2+1, IFFT will panic.
+// If dst is nil, a new slice is allocated and returned. If dst is not nil and
+// the length of dst does not equal the length of coeff, IFFT will panic.
+func (t *FFT) IFFT(dst []float64, coeff []complex128) []float64 {
+	if len(coeff) != t.Len()/2+1 {
+		panic("fourier: coefficients length mismatch")
+	}
+	if dst == nil {
+		dst = make([]float64, t.Len())
+	} else if len(dst) != t.Len() {
+		panic("fourier: destination length mismatch")
+	}
+	dst[0] = real(coeff[0])
+	if len(dst) < 2 {
+		return dst
+	}
+	nf := coeff[len(coeff)-1]
+	if len(dst)%2 == 1 {
+		dst[len(dst)-2] = real(nf)
+		dst[len(dst)-1] = imag(nf)
+	} else {
+		dst[len(dst)-1] = real(nf)
+	}
+
+	for i, cv := range coeff[1 : len(coeff)-1] {
+		dst[2*i+1] = real(cv)
+		dst[2*i+2] = imag(cv)
+	}
+	rfftb(len(dst), dst, t.work, t.ifac[:])
+	return dst
+}
+
+// Freq returns the relative frequency center for coefficient i.
+// Freq will panic if i is negative or greater than or equal to t.Len().
+func (t *FFT) Freq(i int) float64 {
+	if i < 0 || t.Len() <= i {
+		panic("fourier: index out of range")
+	}
+	step := 1 / float64(t.Len())
+	return step * float64(i)
+}
+
+// CmplxFFT implements Fast Fourier Transform and its inverse for real sequences.
+type CmplxFFT struct {
+	work []float64
+	ifac [15]int
+
+	// real temporarily store complex data as
+	// pairs of real values to allow passing to
+	// the backing code. The length of real
+	// must always be half the length of work.
+	real []float64
+}
+
+// NewCmplxFFT returns an CmplxFFT initialized for work on sequences of length n.
+func NewCmplxFFT(n int) *CmplxFFT {
+	var t CmplxFFT
+	t.Reset(n)
+	return &t
+}
+
+// Len returns the length of the acceptable input.
+func (t *CmplxFFT) Len() int { return len(t.work) / 4 }
+
+// Reset reinitializes the FFT for work on sequences of length n.
+func (t *CmplxFFT) Reset(n int) {
+	if 4*n <= cap(t.work) {
+		t.work = t.work[:4*n]
+		t.real = t.real[:2*n]
+	} else {
+		t.work = make([]float64, 4*n)
+		t.real = make([]float64, 2*n)
+	}
+	cffti(n, t.work, t.ifac[:])
+}
+
+// FFT computes the Fourier coefficients of a complex input sequence,
+// seq, placing the result in dst and returning it. This transform is
+// unnormalized since a call to FFT followed by a call of IFFT will
+// multiply the input sequence by the length of the sequence.
+//
+// If the length of seq is not t.Len(), FFT will panic.
+// If dst is nil, a new slice is allocated and returned. If dst is not nil and
+// the length of dst does not equal the length of seq, FFT will panic.
+// It is safe to use the same slice for dst and seq.
```

Comment:
Is it not safe for other calls? It's weird that this one specifically says it but the others don't.

Reply:
This is the only call that can possibly have `dst` and `seq` be the same slice since the real FFT calls don't have matching types for these two parameters.